### PR TITLE
 machines: replace os-vendor/os dropdowns with TypeAheadSelect component from PF-react

### PR DESCRIPTION
**Pull request by KKoukiou**
os-not-found
fedora-filter
os-spinner

### CODE REVIEW

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx [hunk 1]
```diff
@@ -124,8 +121,8 @@ function validateParams(vmParams) {
     // If we select installation media from URL force the user to select
     // OS, since virt-install will not detect the OS, in case we don't choose
     // to start the guest immediately.
-    if (vmParams.vendor == NOT_SPECIFIED && vmParams.sourceType == URL_SOURCE && !vmParams.startVm)
-        validationFailed['vendor'] = _("You need to select the most closely matching OS vendor and Operating System");
+    if ((vmParams.os == undefined || vmParams.os.name == 'Other OS') && vmParams.sourceType == URL_SOURCE && !vmParams.startVm)
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: I thought the previous commit just dropped "Other OS", why is it coming back here?

File: test/avocado/testlib_avocado/machineslib.py [hunk 2]
```diff
@@ -236,11 +236,8 @@ def create_vm_by_ui(self,
             if element:
                 self.click(element)
 
-        if os_vender != 'unspecified':
-            self.select_by_text(self.wait_css('#vendor-select'))
-
-        if os_vender != 'unspecified' and os != 'other':
-            self.select_by_text(self.wait_css('#system-select'), os)
+        if os != 'other':
```
Comment (martinpitt) [inline test/avocado/testlib_avocado/machineslib.py]: Same question -- I thought "other" was a cockpit-ism that got removed in the previous commit; is it an actual valid value in libosinfo/libvirt?

**Comment by martinpitt** [comment]
A single list will mean that it's harder to operate with just the mouse, as it will be a looong merged list, right? Is that an a11y concern?

(I didn't yet test this, waiting for the rebase)

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: The commit that drops the 'Other OS' comes after this one,
in the first commit you still can find the 'Other OS' on the list of available OSes, and we should not allow users to choose it when the media is coming from URL.

File: test/avocado/testlib_avocado/machineslib.py (hunk 2 above)
Comment (KKoukiou) [reply test/avocado/testlib_avocado/machineslib.py]: In this selenium test, they assume that the default os is other, see https://github.com/cockpit-project/cockpit/blob/master/test/avocado/testlib_avocado/machineslib.py#L205
In this case they don't set anything in the dialog which in our case works fine.
I think it will be cleaner to change the other to None in the selenium test.

**Comment by KKoukiou** [comment]
> A single list will mean that it's harder to operate with just the mouse, as it will be a looong merged list, right? Is that an a11y concern?
> 
> (I didn't yet test this, waiting for the rebase)

On the other hand you are able to filter results by typing some substring of the desired distro. 
Please test, and if you still don't like it let's drop this.

**Comment by martinpitt** [comment]
> On the other hand you are able to filter results by typing some substring of the desired distro.

Right, I personally like this a lot, but I'm a very keyboard-affine person. I was just wondering what that means for mouse users.

Also, right now I can type arbitrary strings into that box, like "foo", and there is no error message. It should only be allowed to select an existing entry.

File: test/avocado/testlib_avocado/machineslib.py (hunk 2 above)
Comment (KKoukiou) [reply test/avocado/testlib_avocado/machineslib.py]: > > On the other hand you are able to filter results by typing some substring of the desired distro.
> 
> Right, I personally like this a lot, but I'm a very keyboard-affine person. I was just wondering what that means for mouse users.
> 
> Also, right now I can type arbitrary strings into that box, like "foo", and there is no error message. It should only be allowed to select an existing entry.

Oh good catch, it gives and 'No matches found' but then the dialog gets the value even if it's nonsense. Good catch.

Comment (KKoukiou) [reply test/avocado/testlib_avocado/machineslib.py]: So this seems to be the default behavior of the component, see https://tinyurl.com/y22ckuym
And it's not overwritable,
I thnk I will be able to workaround it in a hacky way, since it's really weird to be able to submit the dialog with false data.

**Comment by garrett** [comment]
This is problematic as-is.

I discussed this with @andreasn and have an idea on how to incorporate this into a working design which would allow:
- automatic matching
- type to search
- browsability

It would basically be two side-by-side components in the dialog.

I don't have time right now to work on a proper mockup, but I'm adding a table-based sketch on what I have in mind. The first column is the label. The second is a select with several options. The third is either a text string for the auto-detect, an input and auto-complete list for search, or a second dropdown for OS version (when browsing based on vendor on the first dropdown)

Label | Primary (Dropdown)  | Secondary (String / Input / Dropdown)  
-- | -- | --
  |   |  
Operating System | Auto-detect | Determining OS…
  |   | Fedora 30
  |   |  
  | Search | [_________]
  |   | (list of matches when search field is used)
  |   |  
  | Red Hat | Supported
  |   | - Red Hat Enterprise Linux 8
  |   | - Red Hat Enterprise Linux 7
  |   | Extended Support
  |   | - Red Hat Enterprise Linux 6
  |   | - Red Hat Enterprise Linux 5
  |   | End of Life
  |   | - Red Hat Enterprise Linux 4
  |   | - Red Hat Enterprise Linux 3
  |   | - Red Hat Enterprise Linux 2
  |   | - Red Hat Enterprise Linux 1
  |   |  
  | Fedora | Supported
  |   | - Fedora 30
  |   | - Fedora Silverblue 30
  |   | - Fedora 29
  |   | - Fedora Silverblue 29
  |   | End of Life
  |   | - Fedora 28
  |   | - Fedora Silverblue 28
  |   |  
  | … | …

It would default to auto-detection. If that doesn't work, the option becomes disabled and then the next selection option should be search, which has the widget from this PR. People can optionally browse by drilling down with vendors and distribution versions, which are then grouped by supportability.

Note: One thing I didn't mock-up is showing supportability in the search or even auto-detection... but this should be shown somehow.

**Comment by andreasn** [comment]
I agree that this becomes very keyboard focused, and there need to be some way to manually select an OS, without knowing exactly what to type.
If we want to keep it to just one field, it would need to be something like this:
https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html

Just showing the entire list straight up and down would end up being _quite_ a long list.
I wonder how long the list would become if we just listed OSes that are not EOL, and then we'll have some way of revealing the full list of the OSes (also EOLed ones)
Is there an easy way to query libosinfo for just  the supported OSes easily via the CLI?

**Comment by andreasn** [comment]
But if we go for the 2-field solution that @garrett proposes, do we really need "Auto-detect" as a dropdown? I think we could just show the right entry in the list and not make a big fuzz about it (just like how it works now), then you would be able to correct that selection, or go for the Search action.

**Comment by andreasn** [comment]
Relevant info, [recognition vs recall](https://uxmyths.com/post/717755413/myth-search-will-solve-a-websites-navigation-problems), see bullet point 3.

**Comment by garrett** [comment]
> do we really need "Auto-detect" as a dropdown [...]?

How long does it take to check?

We _could_ have a placeholder message and then replace it with a widget similar what I described if it fails to detect (except without autodetect). If it doesn't take long. And if it doesn't get it wrong.

**Comment by KKoukiou** [comment]
> I agree that this becomes very keyboard focused, and there need to be some way to manually select an OS, without knowing exactly what to type.
> If we want to keep it to just one field, it would need to be something like this:
> https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html

Which of the mentioned examples you mean?

> 
> Just showing the entire list straight up and down would end up being _quite_ a long list.

It's long, filtering with substrings helps only in this case.
Of course the component implements paging, the list is not presented to the user all at once.

> I wonder how long the list would become if we just listed OSes that are not EOL, and then we'll have some way of revealing the full list of the OSes (also EOLed ones)

I don't think this component allows such logic to be implemented.
Keep in mind that I am using an patternfly component, the flexibility here is not so big.
I can only pass to it the list of entries that it's supposed to present.

> Is there an easy way to query libosinfo for just the supported OSes easily via the CLI?

It's possible to query osinfo-db for EOF for each OS.

**Comment by KKoukiou** [comment]
> > do we really need "Auto-detect" as a dropdown [...]?
> 
> How long does it take to check?

It's a matter of 1-2 seconds after inserting the installation source.
There is a spinner shown when the component is trying to detect the OS.
> 
> We _could_ have a placeholder message and then replace it with a widget similar what I described if it fails to detect (except without autodetect). If it doesn't take long. And if it doesn't get it wrong.

It can't really get it wrong, it can only not get it at all.
I can replace the placeholder when it's searching. That's not a problem.

**Comment by andreasn** [comment]
> I agree that this becomes very keyboard focused, and there need to be some way to manually select an OS, without knowing exactly what to type.

This was false actually. I didn't have libosinfo installed and didn't realize I needed to.
So selecting the input field will give you a list of OSes. I couldn't quite figure out how this list is ordered, and it seems a bit random to me, because I get a dozen versions of FreeBSD at the top, followed by Fedora, then Ubuntu and then a dozen DragonflyBSD. So it seems I am able to browse for the OSes, but the sorting doesn't make sense to me.

**Comment by KKoukiou** [comment]
> > I agree that this becomes very keyboard focused, and there need to be some way to manually select an OS, without knowing exactly what to type.
> 
> This was false actually. I didn't have libosinfo installed and didn't realize I needed to.


You don't have libosinfo installed because you built cockpit-machines from sources and you don't have its dependencies.
For normal users it will be there.

> So selecting the input field will give you a list of OSes. I couldn't quite figure out how this list is ordered, and it seems a bit random to me, because I get a dozen versions of FreeBSD at the top, followed by Fedora, then Ubuntu and then a dozen DragonflyBSD. So it seems I am able to browse for the OSes, but the sorting doesn't make sense to me.

Probably the order get's messed up somewhere indeed.
I will adjust to preserve the order that osinfo-query gives to us
 osinfo-query os  --fields=name,eol-date | tail -n +3 | sed -e 's/\s*|\s*/|/g; s/^\s*//g; s/\s*$//g'

**Comment by andreasn** [comment]
It seems I can also write in anything in freeform now. I tried writing "lennart" and it happily accepted that. Should we check that whatever is written actually matches with a real OS name? Otherwise there is the risk you'll get weird results because someone mistypes "Windows Server 2012" as "Qindows Aerver 2012" or stuff like that and the installation fails due to not setting some flag right.

**Comment by KKoukiou** [comment]
> It seems I can also write in anything in freeform now. I tried writing "lennart" and it happily accepted that. Should we check that whatever is written actually matches with a real OS name? Otherwise there is the risk you'll get weird results because someone mistypes "Windows Server 2012" as "Qindows Aerver 2012" or stuff like that and the installation fails due to not setting some flag right.

Yes, this issue is already mentioned here https://github.com/cockpit-project/cockpit/pull/12152#issuecomment-505792721, I am going to address it if we agree to use this approach.

**Comment by KKoukiou** [comment]
@garrett @andreasn I sorted the list according to the OS name, and within the same vendor the OSes are  sorted in drecreasing order according to the release date (same as before)
I also added a warning, in case a user chooses an OS that has reached its EOL.
See:
eol-date

@andreasn  I fixed you comment, to clear the choice onblur in case it's invalid.

Now, I was thinking about the pros and cons of having a single list again vs split vendor-OS and I realized another advantage of this design.
The problem is that sometimes the vendor names don't match the OS name. See
Apple -> MacOS
Microsoft Corp -> Windows
Canonical -> Ubuntu etc

So when a user wants to choose Ubuntu as an OS, he needed to remember the name of the organization behind it as well.
So if we have smaller list currently, but we really require from the user information which is not trivial.

Other reason that I really prefer the implemented here design is that is cleanly encapsulates the empty choice.
In fact, currently we don't force the user to choose an OS, and we have the default Unspecified OS in the dropdown.
However, once we introduce unattended installations the OS should be mandatory field, since the user will just insert the OS, and the tools we use will determine the installation source, recommended memory, disk size etc.
In that case of course, auto-detection will not be used, since no source field would be present either.
For that case, I really prefer to have a component, that cleanly shows that it's not filled yet.
Just for clarity what I am talking about here is a draft of the UI I am implementing now for unattended installations.
unattended

The default OS selection is invalid for unattended installations.

**Comment by garrett** [comment]
Does the auto-detection check the filename of ISOs? That seems like a quick fist-pass autodetection. Ubuntu ISOs have Ubuntu in the name, so does Fedora, so does openSUSE, etc.

We could even pass in the filename (or part of the filename) as a string to match against, similar to the dropdown.

Everything I checked has the format Distroname-spin-architecture-version.iso — it would, at least, be a good, quick first pass to get an immediate match while libosinfo does its thing.... and probably be correct 98% of the time. So we could do a two-pass autodetection for both speed and accuracy.

**Comment by KKoukiou** [comment]
> Does the auto-detection check the filename of ISOs? That seems like a quick fist-pass autodetection. Ubuntu ISOs have Ubuntu in the name, so does Fedora, so does openSUSE, etc.
> 
> We could even pass in the filename (or part of the filename) as a string to match against, similar to the dropdown.
> 
> Everything I checked has the format Distroname-spin-architecture-version.iso — it would, at least, be a good, quick first pass to get an immediate match while libosinfo does its thing.... and probably be correct 98% of the time. So we could do a two-pass autodetection for both speed and accuracy.

I don't do the auto detection by myself. I heavily believe we should rely on external tools that do it better (see osinfo-detect) which is what we do right now.

**Comment by KKoukiou** [comment]
@garrett see this screeshot, it's browsable as well
browseable

**Comment by KKoukiou** [comment]
@garrett so our plan to move to the pf4 equivalent component now is not feasible because of all the issues we found out. Currently the only problem of this PF3 component if I understand correctly is that it's missing the caret to indicate that the list is expandable with mouse.
I don't know how to place such arrow inside the text box (relative position with css?)
But if we show an arrow are you ok with this solution?

File: test/avocado/testlib_avocado/machineslib.py (hunk 2 above)
Comment (KKoukiou) [reply test/avocado/testlib_avocado/machineslib.py]: Done

**Comment by garrett** [comment]
Thanks for your effort on this, but I don't approve.

Let's pretend we're an end-user and want to create a VM for RHEL 8.

## Browsing

1. Click the dropdown, scroll down... and it's not listed. There's just a bunch of "ALT", "Alpine", "DragonFly"...
2. Keep scrolling to the end, click "load more"
3. Keep scrolling, still no Red Hat Enterprise Linux.
4. Give up.

## Searching

1. Type "rhel". No matches.
2. Type "rh". No matches.
3. Type "redhat". No matches.
4. Type "Red Hat 8". No Matches.
5. Give up.


---

Sorry, but right now, it's a much worse experience than what we already have. And I don't know if this is even accessible. (It probably isn't.)

Native selects are accessible.

---

What I'd like to see:

- First class typing _and_ browsing
- Prune the list down to supported releases in browse mode
- Have some style of fuzzy matching — it's enough to do char*char*char matches like on GitHub (when you hit "t" at the repo browsing) or control-p or control-t in many IDEs. ("rhel8" would match "*R*ed *H*at *E*nterprise *L*inux *8*.0", for example)
- Accessibility support for keyboards and screen readers in any new widget we bring in or create

If PatternFly 3 and 4 are both horribly broken in their own ways, then I guess we'll have to create our own widget. Perhaps something like [ARIA 1.1 Combobox with Listbox Popup Examples: Example 3](https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html) might be a good starting point? I would be happy to make sure it's styled accordingly.

**Comment by KKoukiou** [comment]
> Thanks for your effort on this, but I don't approve.
> 
> Let's pretend we're an end-user and want to create a VM for RHEL 8.

Please who are installating RHEL guest will fill first the installation source option and the OS will be autocompleted. But of course this is not enough: See comments inline.

> ## Browsing
> 
>     1. Click the dropdown, scroll down... and it's not listed. There's just a bunch of "ALT", "Alpine", "DragonFly"...
> 
>     2. Keep scrolling to the end, click "load more"
> 
>     3. Keep scrolling, still no Red Hat Enterprise Linux.
> 
>     4. Give up.

The list is long indeed and alphabetically that OS is close to the end. However I agree with not showing the EOL OSes at all in general. I thought of it and I think it's useless. Even if people want install some non supported OS can leave the field either blank or choose some close match, which will be just fine.
> 
> 
> ## Searching
> 
>     1. Type "rhel". No matches.
> 
>     2. Type "rh". No matches.
> 
>     3. Type "redhat". No matches.

When typing Red you will see the displayed matches, the lisst renders super fast. 
> 
>     4. Type "Red Hat 8". No Matches.

Same here, Red Hat will show all Red Hat provided OSes.
Let's not exagerate regarding the filtering.
Of course some smarter filtering function could exist but I don't think it should be considered blocking issue.
 
> 


>     5. Give up.
> 
> 
> Sorry, but right now, it's a much worse experience than what we already have. And I don't know if this is even accessible. (It probably isn't.)

I actually tested with ChromeVox and it's accessible and behaves very nicely with screen reader.

I also believe that for non mouse only users the current experience is not worse, quite the opposite. I wouldn't have bothered to change it otherwise. I am really frustrated with having to select two input fields just for something that I can type in 2 seconds.

Of course for mouse users, I want to do everything I can to make it good experience. 
> 
> Native selects are accessible.
> 
> What I'd like to see:
> 
>     * First class typing _and_ browsing
> 
>     * Prune the list down to supported releases in browse mode

We can completety remove EOL OSes. That's good point raised here.

> 
>     * Have some style of fuzzy matching — it's enough to do char_char_char matches like on GitHub (when you hit "t" at the repo browsing) or control-p or control-t in many IDEs. ("rhel8" would match "_R_ed _H_at _E_nterprise _L_inux _8_.0", for example)

I think this is doable but I would rather not implement it within the scope if this PR.

> 
>     * Accessibility support for keyboards and screen readers in any new widget we bring in or create
> 

Already accessible.

> 
> If PatternFly 3 and 4 are both horribly broken in their own ways, then I guess we'll have to create our own widget. Perhaps something like [ARIA 1.1 Combobox with Listbox Popup Examples: Example 3](https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html) might be a good starting point? I would be happy to make sure it's styled accordingly.

The only thing broken with PF3 typeahead select component is that it doesn't show the caret. PF4 is horribly broken indeed.
I really believe we should use PF components as much as possible when they don't have real importan tissues than implementing our own. And for minor issues we should open bugs (even mark them as blockers) and let them fix them, and we should fix them directly in PF that creating our own components.

**Comment by KKoukiou** [comment]
@garrett as we agreed yesterday I trimmed down the list of OSes to contains the ones with EOL date till max one year ago. And release date max 3 years ago.

I also adjusted the filter function of the typeahead componen to perform filtering by ID as well as name.
So now you can type Red Hat, rhel, rhel8 and all will give the correct results.

**Comment by KKoukiou** [comment]
Adding needs work so that @garrett will add the missing caret.

**Comment by garrett** [comment]
I have good new and bad news.

First, the good:
- I fixed the caret locally!

Now, the bad:
1. You have to rebase this branch on top of master.
   - My fix depends on code that was recently merged. There's no way around this.
2. Keyboard selection is broken.
   - The colors are black on blue and it's almost impossible to read.
   - And I have no clue how to find out what to style. (I tried obvious things already.)
   - I can't make heads or tails of the weird react code in https://github.com/patternfly/patternfly-react/tree/master/packages/patternfly-3/patternfly-react/src/components/TypeAheadSelect either. Looking at the sources just does not help here.
3. Searching for "rhel" (or even "rh") still doesn't work... :wink:

**Comment by KKoukiou** [comment]
> I have good new and bad news.
> 
> First, the good:
> 
>     * I fixed the caret locally!
> 
> 
> Now, the bad:
> 
>     1. You have to rebase this branch on top of master.
>        
No problem, rebased.

>        * My fix depends on code that was recently merged. There's no way around this.
> 
>     2. Keyboard selection is broken.
>        
>        * The colors are black on blue and it's almost impossible to read.
>        * And I have no clue how to find out what to style. (I tried obvious things already.)
>        * I can't make heads or tails of the weird react code in https://github.com/patternfly/patternfly-react/tree/master/packages/patternfly-3/patternfly-react/src/components/TypeAheadSelect either. Looking at the sources just does not help here.

Let me try to find the variable that defined the background color for the dropdown entries.
> 
>     3. Searching for "rhel" (or even "rh") still doesn't work... wink

It works for me, maybe you are missing dependencies like @andreasn  above?

**Comment by garrett** [comment]
I pushed a commit that adds the dropdown caret icon.

I _don't_ have a fix for the dropdown's keyboard focus colors being wrong (black on dark blue). I need to know what the HTML looks like. When I try to select it in dev tools, it disappears, and I can't discover it. I haven't been able to find out what classes or elements it uses either. Can you help me find this out? If so, I can make a fix for this too.

**Comment by KKoukiou** [comment]
> I pushed a commit that adds the dropdown caret icon.
> 
> I _don't_ have a fix for the dropdown's keyboard focus colors being wrong (black on dark blue). I need to know what the HTML looks like. When I try to select it in dev tools, it disappears, and I can't discover it. I haven't been able to find out what classes or elements it uses either. Can you help me find this out? If so, I can make a fix for this too.

They reason it disappears when you try to right click on it is because of an ugly hack I have there, to make it work the way we want.
The default behaviour of the component is to keep the string text in the input, even if the user does not select any of the available options. 
For example, the user might write 'katerina' in the input, move away and the string 'katerina' would remain there in the default behaviour. All of the people who reviewed this, said that this is not desired, so in order to clear the input text, I do force-reload the component when onBlur event gets fired.
That's what happens when you right click, the component reloads.

TBH I don't like this hack, but I was not able to to it in a nicer way.
To stop temporarily getting this behaviour you can remove the onBlur callback from https://github.com/cockpit-project/cockpit/pull/12152/commits/7af220ca7ed0639d90fdb1ae5e6d961d75d8a9bf#diff-ca27be13c337ff78870b5a86dbdea08eR330

**Comment by garrett** [comment]
OK, I've made another commit to override PF3's very specific sets of !important for the active state by also being specific and using !important. :roll_eyes: This fixes the dropdown keyboard colors. (Please keep these PF override fixes as separate commits.)

In testing it, I noticed that hitting ESC dismisses the dialog. Can you trap hitting ESC to only affect the dropdown when it has keyboard focus please?

Basically, on the typing event handler, check for ESC being pressed and use `event.stopPropagation();` to stop it from bubbling up. (Probably.)

**Comment by garrett** [comment]
Also, why are there nested levels of classless `<span>`s? Is this from PF3 React not using React.Fragment here or something? Or is it in our code?

**Comment by KKoukiou** [comment]
> Also, why are there nested levels of classless `<span>`s? Is this from PF3 React not using React.Fragment here or something? Or is it in our code?

These are all brought by either PF-react or even bootstrap implement ion of this component that PF-3 is using. I am afraid we can't affect it.

**Comment by garrett** [comment]
@KKoukiou: Ugh. OK. At least _we_ know better. :smile:

Perhaps the React widgets were written with an older version of React in mind (one without React.Fragment)?

At least, in this case, it probably doesn't break anything, except a tiny, most likely unnoticeable amount of performance loss from a bit more complex DOM.

**Comment by KKoukiou** [comment]
> OK, I've made another commit to override PF3's very specific sets of !important for the active state by also being specific and using !important. roll_eyes This fixes the dropdown keyboard colors. (Please keep these PF override fixes as separate commits.)
> 
> In testing it, I noticed that hitting ESC dismisses the dialog. Can you trap hitting ESC to only affect the dropdown when it has keyboard focus please?
> 
> Basically, on the typing event handler, check for ESC being pressed and use `event.stopPropagation();` to stop it from bubbling up. (Probably.)

Done.

**Comment by marusak** [comment]
Just mentioning that this also needs https://github.com/cockpit-project/cockpit/pull/12331/commits/156ddeb8a4965d7a04a72618606cc0250dee70b0 from #12331 otherwise the placeholder does not align with the real text.

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx [hunk 3]
```diff
@@ -279,66 +280,75 @@ const SourceRow = ({ source, sourceType, networks, nodeDevices, providerName, on
...
-        vendorSelectEntries.push((<Select.SelectDivider key='divider' />));
-    }
+class OSRow extends React.Component {
+    constructor(props) {
+        super(props);
+        const IGNORE_VENDORS = ['ALTLinux', 'Mandriva', 'GNOME Project'];
+        const osInfoListExt = this.props.osInfoList
+                .filter(os => !IGNORE_VENDORS.find(vendor => vendor == os.vendor))
+                .map(os => correctSpecialCases(os))
+                .filter(os => filterReleaseEolDates(os))
```
Comment (marusak) [inline pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: `filter->map->filter` could this be done with `map->filter` or just one `reduce`?

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx [hunk 4]
```diff
@@ -279,66 +280,75 @@ const SourceRow = ({ source, sourceType, networks, nodeDevices, providerName, on
...
-
-    const osEntries = (
-        vendorMap[vendor]
-                .map(os => (
-                    <Select.SelectEntry data={os.shortId} key={os.shortId}>
-                        {getOSStringRepresentation(os)}
-                    </Select.SelectEntry>))
-    );
+        this.state = {
+            typeAheadKey: Math.random(),
```
Comment (marusak) [inline pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: Using random as key is not wrong but not optimal. Is every item going to be in the list just once, if so, then you can construct some string like `shortId+name+version`.

Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: This is done on the purpose actually. See onBlur callback on the TypeAheadSelect.

The thing is that, by default that component, when you onfocus, it will keep in the input text whatever you typed, regardless it's matching with one of the preselected elemements.
I couldn't set the value of the input field, (this make the component controlled)
So I decided to reset the whole component when onBlur event arrives.
The only clear way to reset a react component, it to change it's key.
I will be glad it you have some suggestion on how to ger rid of this ugly hack of course!

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: Done.

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx (hunk 4 above)
Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: > Small nitpicks.
> Also there is typo in one commit title `machines: OS now are filtered by shortId as well the OS diplay name` and one of @garrett commit could be squashed.
> 
> And I think that you know, but tests still don't seems to get green :(

Test issue hopefully fixed with repost.

**Comment by KKoukiou** [comment]
> Small nitpicks.
> Also there is typo in one commit title `machines: OS now are filtered by shortId as well the OS diplay name` and one of @garrett commit could be squashed.

Which commit you mean? global PF fixes should be in separate commits.

> 
> And I think that you know, but tests still don't seems to get green :(
 
tests hopefully fixed now.

**Comment by KKoukiou** [comment]
@garrett one more issue here :( The spinner is on the right exactly where the caret is placed now. Can we move it a bit on the left or something?

**Comment by garrett** [comment]
@KKoukiou: How do you trigger a spinner?! (I really don't see how here locally.)

**Comment by KKoukiou** [comment]
> @KKoukiou: How do you trigger a spinner?! (I really don't see how here locally.)

Just trigger the auto detection by placing an installation URL in the source, ex: https://download.fedoraproject.org/pub/fedora/linux/releases/29/Server/x86_64/os

**Comment by garrett** [comment]
@KKoukiou: What's the expected behavior? Should the dropdown icon disappear while there's a spinner, or should it go to the left?

**Comment by KKoukiou** [comment]
There is one quite persistent test failure in rhel-8-1 here https://209.132.184.41:8493/logs/pull-12152-20190719-143737-7029288e-verify-rhel-8-1/log.html 
I doubt that it's related, but I'd rather block this one on fixing this test, this added needswork.

**Comment by marusak** [comment]
Needs to rebase to master since #12367 changed tests names

**Comment by KKoukiou** [comment]
Blocking on https://github.com/cockpit-project/cockpit/pull/12406
since rhel-8-1 fails on exactly that bug.

**Comment by KKoukiou** [comment]
In case we include it in release notes: 
Screen Shot 2019-08-01 at 16 27 29

**Comment by garrett** [comment]
Here's a cropped (with 25px margin) and optimized version of the screenshot:

62301673-50bb3780-b479-11e9-9356-ea01e65f0c86-cropped (copy)
